crates/wasmtime/src/runtime/vm/memory.rs

```diff
@@ -488,14 +522,28 @@ impl LocalMemory {
                 // `RuntimeLinearMemory::byte_size` is not a multiple of the host page
                 // size. See https://github.com/bytecodealliance/wasmtime/issues/9660.
                 if let Ok(byte_size) = HostAlignedByteCount::new(alloc.byte_size()) {
-                    let mut slot = MemoryImageSlot::create(
-                        alloc.base_ptr().cast(),
-                        byte_size,
-                        alloc.byte_capacity(),
-                    );
-                    // On drop, we will unmap our mmap'd range that this slot was
-                    // mapped on top of, so there is no need for the slot to wipe
-                    // it with an anonymous mapping first.
+                    // memory_image is CoW-based so it is expected to be backed
+                    // by an mmap.
+                    let mmap_base = match alloc.base() {
+                        MemoryBase::Mmap(offset) => offset,
+                        MemoryBase::Raw { .. } => {
+                            unreachable!("memory_image is Some only for mmap-based memories")
+                        }
+                    };
+
+                    let mut slot =
+                        MemoryImageSlot::create(mmap_base, byte_size, alloc.byte_capacity());
+                    // On drop, we will unmap our mmap'd range that this slot
+                    // was mapped on top of, so there is no need for the slot to
+                    // wipe it with an anonymous mapping first.
+                    //
+                    // Note that this code would be incorrect if clear-on-drop
+                    // were enabled. That's because in the struct definition,
+                    // `memory_image` above is listed after `alloc`. Rust drops
+                    // fields in the order they're defined, so `memory_image`
+                    // would be dropped after `alloc`. If clear-on-drop were
+                    // enabled, `memory_image` would end up remapping a bunch of
+                    // memory after it was unmapped by the `alloc` drop.
```

Comment:
This was a subtle dependency that took me a bit to understand.

Reply:
(And this is only dangerous when `alloc` represents an owned memory allocation. It's fine if `alloc` is borrowed. This is a subtle point.)

Reply:
Do you think it'd perhaps be better to remove this `no_clear_on_drop` and instead perform it in `Drop for LocalMemory`? (where that could `.take()` and explicitly do the image stuff before dropping the mmap)

Reply:
Oh that's a good idea, similar to `MemoryPool`. I think it's correct.

Reply:
Oh actually there's another subtle issue here, I think if `slot.instantiate` below fails, we should clear on drop (that's part of the point of clear-on-drop, right?) So I think it would be best to make this change in a followup since it's a behavior change, not a refactor.

Reply:
I've expanded this comment -- it would definitely be good to move this to the destructor, but it would be a behavior change so it should be a followup. This PR itself should not contain any behavior changes.

Reply:
Oh, and adding a `Drop` impl identified the problem with `unwrap_static_image`, haha, since you can't move out of a type when there's a `Drop` impl.

Reply:
Ah ok that all makes sense to me yeah, thanks for checking!